src/libos/src/signal/do_sigreturn.rs

```diff
@@ -393,47 +379,47 @@ impl Stack {
     }
 }
 
-thread_local! {
-    static PRE_UCONTEXTS: RefCell<CpuContextStack> = Default::default();
+async_rt::task_local! {
+    static SIGNAL_CONTEXT: RefCell<SignalContext> = RefCell::new(SignalContext::new());
 }
 
-#[derive(Debug, Default)]
-struct CpuContextStack {
-    stack: [Option<*mut ucontext_t>; 32],
-    count: usize,
+struct SignalContext {
+    context: *mut ucontext_t,
 }
 
-impl CpuContextStack {
+impl SignalContext {
     pub fn new() -> Self {
-        Default::default()
-    }
-
-    pub fn full(&self) -> bool {
-        self.count == self.stack.len()
+        SignalContext {
+            context: ptr::null_mut(),
+        }
     }
 
-    pub fn empty(&self) -> bool {
-        self.count == 0
-    }
+    pub fn push(&mut self, cpu_context: *mut ucontext_t) {
```

Comment:
This method should be marked `unsafe`. This is a common problem in my review experience. Many developers do not have a clear understanding when `unsafe` should be used.

Put simply, mark a function `unsafe` if the function requires some conditions that cannot be checked by the Rust compiler to ensure memory safety. Apparently, this function falls in the category as it has an implicit requirement that the `cpu_context` pointer is valid, pointing to valid memory region and having a proper alignment. Otherwise, the Rust code may have undefined behaviors.